src/lib.rs

```diff
@@ -339,7 +339,7 @@ mod test {
     #[test]
     fn check_array_ref_5() {
         fn f(data: Vec<u8>, offset: usize) -> quickcheck::TestResult {
-            if data.len() < offset + 5 {
+            if data.len() < 5 || data.len() - 5 < offset {
```

Comment:
**suggestion:** I think this would be clearer with `checked_add`.

My brain hurts just a bit when trying to understand this logical combination, and if I hadn't read your description it wouldn't have been obvious that you were dealing with overflow.
```suggestion
            if data.len() < offset.checked_add(5).unwrap_or_default() {
```
Does this feel more clear to you also? If so, could you make this change to all of these?

Reply:
The suggestion to try to use a `*_add` method with more appropriate behavior rather than plain `+` or `-` for readability is a reasonable one. Unfortunately, it doesn’t look like this suggestion is logically equivalent.

Consider the case where `usize` is 64-bit, `data.len()` is `3` and `offset` is `usize::MAX - 2`, i.e. `0xfffffffffffffffe` or `18446744073709551614`. Then considering the intent of the original `data.len() < offset + 5`, we want the result to be `true`.

- The original expression `data.len() < offset + 5` got this wrong; `0xfffffffffffffffe + 5` wraps around to `3`, so the comparison is `5 < 3`, which is `false`.
- In `data.len() < 5 || data.len() - 5 < offset`, the first condition short-circuits to `true`: if `data.len()` is smaller than `5`, it must be less than the (mathematical, unbounded) sum of `5` and any nonnegative offset. Then the remaining comparison is written in a way that uses the reduction in the domain of `data.len()` due to the first condition to make wrapping impossible.
- In `offset.checked_add(5).unwrap_or_default()`, overflow is detected, so the `Option<usize>` is `None`, and so `.unwrap_or_default()` gives us `0`. The comparison is `5 < 0`, which is `false` – the same flaw as the original code, arrived at in a different way.

I do understand that it’s not obvious at a glance that `if data.len() < 5 || data.len() - 5 < offset` is an overflow-proof version of the idea in `data.len() < offset + 5` – some programmers accustomed to dealing with this in languages like C or C++ will recognize the pattern, but it still requires some thought to recognize the intent. I’ll think about whether there is a correct, obvious-looking way to express this with one of the a `*_add` methods, or whether perhaps it’s better to just add brief explanatory comments to these awkward conditionals.

Reply:
In general, the difficulty in a simple fallback without rearranging the comparison is that when overflow occurs, we still need to evaluate the expression correctly, without discarding information from the operands, and one way or the other this tends to lead to needing to express it twice in perhaps slightly unnatural ways. But I’ll keep thinking about clever alternatives.

Reply:
Even considering other methods like `.saturating_sub()`, I can’t think of a way to do this with `.<foo>_add.unwrap_or_<bar>()`. Anything I can think of has to have some kind of branch leading to the final Boolean result instead. It’s possible I am missing a trick, but I think it’s unlikely I’m missing one that’s more legible than the original suggestion.

For now, I’m just going to add comments to these four expressions, but if you have any other suggestions, I’m happy to try to prove/disprove their logical equivalence.